bpzpipe/BPZPipe2.py

```diff
@@ -0,0 +1,351 @@
+        group.create_dataset("mu_1p", (nobj,), dtype='f4')
+        group.create_dataset("mu_1m", (nobj,), dtype='f4')
+        group.create_dataset("mu_2p", (nobj,), dtype='f4')
+        group.create_dataset("mu_2m", (nobj,), dtype='f4')
+
+        # One processor writes the redshift axis to output.
+        if self.rank==0:
+            group['z'][:] = z
+
+        return f
+
+    def load_templates(self):
+        from useful_py3 import get_str, get_data, match_resol
+
+
+        # The reshift range we will evaluate on
+        zmin = self.config['zmin']
+        zmax = self.config['zmax']
+        dz = self.config['dz']
+        z = np.arange(zmin, zmax + dz, dz)
+
+
+        bpz_path = self.config['path_to_bpz']
+        columns_file = '/Users/jaz/src/BPZpipe/test/CSDC2114_test.columns'
+        ignore_rows =['M_0','OTHER','ID','Z_S']
+        filters = [f for f in get_str(columns_file, 0) if f not in ignore_rows]
+
+        m0_index = 'ugrizy'.index('i')
+        spectra_file = os.path.join(bpz_path, self.config['spectra_file'])
+        spectra = [s[:-4] for s in get_str(spectra_file)]
+        ab_dir = os.path.join(bpz_path, self.config['ab_dir'])
+
+        nt = len(spectra)
+        nf = len(filters)
+        nz = len(z)
+        flux_templates = np.zeros((nz,nt,nf))
+
+        for i,s in enumerate(spectra):
+            for j,f in enumerate(filters):
+                model = f"{s}.{f}.AB"
+                model_path=os.path.join(ab_dir, model)
+                zo, f_mod_0 = get_data(model_path,(0,1))
+                flux_templates[:,i,j] = match_resol(zo, f_mod_0, z)
+
+        return z, flux_templates
+
+    def preprocess_magnitudes(self, data):
+        from bpz_tools_py3 import mag2flux, e_mag2frac
+
+        bands = self.config['bands']
+        suffices = ["", "_1p", "_1m", "_2p", "_2m"]
+
+        # Load the magnitudes
+        zp_errors = np.array(self.config['zp_errors'])
+        zp_frac=e_mag2frac(zp_errors)
+
+        # Only one set of mag errors
+        mag_errs = np.array([data[f'mag_err_{b}_lsst'] for b in bands]).T
+
+        # But many sets of mags, for now
+        for suffix in suffices:
+            # Group the magnitudes and errors into one big array
+            mags = np.array([data[f'mag_{b}_lsst{suffix}'] for b in bands]).T
+
+            # Clip to min mag errors
+            np.clip(mag_errs, self.config['mag_err_min'], 1e10, mag_errs)
+
+            # Convert to pseudo-fluxes
+            flux = 10.0**(-0.4*mags)
+            flux_err = flux * (10.0**(0.4*mag_errs) - 1.0)
+
+            # Check if an object is seen in each band at all.
+            # Fluxes not seen at all are listed as infinity in the input,
+            # so will come out as zero flux and zero flux_err.
+            # Check which is which here, to use with the ZP errors below
+            seen1 = (flux > 0) & (flux_err > 0)
+            seen = np.where(seen1)
+            unseen = np.where(~seen1)
```

Comment:
I would replace this with 
```
nondetect = 99.
nondetect_flux = 10.**(-0.4*nondetect)
unseen = np.isclose(flux,nondetect_flux,atol=nondetectflux*0.5)
flux[unseen] = 0.
flux_err[unseen] = 10.**(-0.4*np.abs(mag_errs[unseen]))
```
or something similar

Reply:
I'm a bit confuse by this line, which I do find in the BPZ code as well:

```
flux_err[unseen] = 10.**(-0.4*np.abs(mag_errs[unseen]))
```
this seems to imply that larger mag errors give smaller flux errors?  Is the negative sign right?

Reply:
Yes, in that we assume that the mag_error column for "unseen" objects does not actually encode a magnitude error, it instead holds the 1 sigma limit for the observation (in units of magnitudes).  The thinking is that if the object is undetected, all you really know for sure is that the flux is somewhere below the noise from the background, and so it assigns an error equal to the 1 sigma background.  So, if the sky background limited you to, say, observing a 27.5 magnitude point source at the 1 sigma level, you would put 27.5 here, whereas if it was a deeper observation you would have e.g. 28.0.  Having a fainter 1 sigma limit should lead to smaller errors, so the minus sign is correct.

It's a confusing convention for sure, I think it was just a way to use the same column space without defining a separate column/variable for some objects that holds the background estimate.